File: nim_edn/edn.nim
## edn.nim — EDN reader producing SExpr values.
##
## Reads the Datomic-style tx-data subset of EDN:
##   vectors [..], lists (..), keywords :ns/name, symbols, ?vars,
##   strings, ints (incl. negative tempids), floats, true/false/nil, `_`.
##
## Maps {..} and sets #{..} are parsed (reader-level) so error messages
## are precise, but the wire rejects them (no consumer yet — see
## docs/tx-protocol.md §4).  A keyword's kwval holds the name WITHOUT the
## leading colon (":person/name" → kwval "person/name"), matching the
## canonical storage form (tx-protocol.md §8).
##
## The reader is fail-loud: malformed input raises EdnError with position.

import std/strutils
import scheme

type
  EdnError* = object of CatchableError

const
  Whitespace = {' ', '\t', '\n', '\r', ','}   # EDN allows comma as whitespace

proc skipWs(input: string; pos: var int) =
  while pos < input.len and input[pos] in Whitespace:
    inc pos

proc parseEdnValue(input: string; pos: var int): SExpr {.raises: [EdnError], gcsafe.}

proc parseEdnString(input: string; pos: var int): SExpr {.raises: [EdnError], gcsafe.} =
  ## pos sits on the opening quote.
  inc pos
  var s = ""
  while pos < input.len:
    let c = input[pos]
    if c == '"':
      inc pos
      return newStr(s)
    if c == '\\':
      if pos + 1 >= input.len:
        raise newException(EdnError, "edn: unterminated escape at end of input")
      let esc = input[pos + 1]
      inc pos, 2
      case esc
      of 'n': s.add '\n'
      of 't': s.add '\t'
      of 'r': s.add '\r'
      of '"': s.add '"'
      of '\\': s.add '\\'
      else:
        raise newException(EdnError, "edn: unsupported escape \\" & esc &
          " at pos " & $pos)
    else:
      s.add c
      inc pos
  raise newException(EdnError, "edn: unterminated string")

proc parseEdnAtom(input: string; pos: var int): SExpr {.raises: [EdnError], gcsafe.} =
  var start = pos
  while pos < input.len and input[pos] notin Whitespace and
        input[pos] notin {'(', ')', '[', ']', '{', '}', '"'}:
    inc pos
  let s = input[start ..< pos]
  if s.len == 0:
    raise newException(EdnError, "edn: empty atom at pos " & $start)
  if s == "nil": return newVoid()
  if s == "true": return newBool(true)
  if s == "false": return newBool(false)
  if s == "_": return newSymbol("_")
  if s.startsWith(':'):
    if s.len == 1:
      raise newException(EdnError, "edn: bare ':' at pos " & $start)
    return newKeyword(s[1 ..^ 1])
  try: return newInt(parseBiggestInt(s))
  except ValueError:
    # Grammar dispatch: int → float → symbol. Not an error path.
    try: return newFloat(parseFloat(s))
    except ValueError: return newSymbol(s)

proc parseEdnColl(input: string; pos: var int; closing: char;
                  what: string): seq[SExpr] {.raises: [EdnError], gcsafe.} =
  ## pos sits just past the opening delimiter.
  while true:
    skipWs(input, pos)
    if pos >= input.len:
      raise newException(EdnError, "edn: unterminated " & what)
    if input[pos] == closing:
      inc pos
      return result
    if input[pos] == '"':
      result.add parseEdnString(input, pos)
    else:
      result.add parseEdnValue(input, pos)

proc parseEdnValue(input: string; pos: var int): SExpr {.raises: [EdnError], gcsafe.} =
  skipWs(input, pos)
  if pos >= input.len:
    raise newException(EdnError, "edn: unexpected end of input")
  case input[pos]
  of '[':
    inc pos
    return newList(parseEdnColl(input, pos, ']', "vector"))
  of '(':
    inc pos
    return newList(parseEdnColl(input, pos, ')', "list"))
  of '{':
    raise newException(EdnError, "edn: maps not supported at pos " & $pos)
  of '#':
    if pos + 1 < input.len and input[pos + 1] == '{':
      raise newException(EdnError, "edn: sets not supported at pos " & $pos)
    raise newException(EdnError, "edn: unsupported dispatch #" & input[pos + 1] &
      " at pos " & $pos)
  of '"':
    return parseEdnString(input, pos)
  else:
    return parseEdnAtom(input, pos)

proc readEdn*(input: string): SExpr {.raises: [EdnError], gcsafe.} =
  ## Read a single EDN value from the whole input.  Trailing content
  ## (other than whitespace) is an error.
  var pos = 0
  result = parseEdnValue(input, pos)
  skipWs(input, pos)
  if pos != input.len:
    raise newException(EdnError, "edn: trailing content at pos " & $pos)

proc readEdnVector*(input: string): seq[SExpr] {.raises: [EdnError], gcsafe.} =
  ## Read a top-level vector `[op op ...]` and return its elements —
  ## the tx-data entry point (docs/tx-protocol.md §3.1).
  var pos = 0
  skipWs(input, pos)
  if pos >= input.len or input[pos] != '[':
    raise newException(EdnError, "edn: expected tx-data vector, got: " &
      (if pos < input.len: $input[pos] else: "<eof>"))
  inc pos
  result = parseEdnColl(input, pos, ']', "tx-data vector")

File: ocaml/lib/edn.ml
(* edn.ml — EDN reader for the Datomic-style tx-data subset, mirroring
   nim_edn/edn.nim:
     vectors [..], lists (..), keywords :ns/name (kwval WITHOUT the
     leading colon), symbols, ?vars, strings, ints (negative tempids),
     floats, true/false/nil, `_`.  Commas are whitespace.  Maps {..} and
     sets #{..} are rejected.  Fail-loud with position. *)

exception Edn_error of string

type t =
  | Int of int64
  | Float of float
  | Str of string
  | Bool of bool
  | Nil
  | Keyword of string
  | Symbol of string
  | List of t list

let is_ws c = c = ' ' || c = '\t' || c = '\n' || c = '\r' || c = ','

let skip_ws (s : string) pos =
  let rec loop pos = if pos < String.length s && is_ws s.[pos] then loop (pos + 1) else pos in
  loop pos

let is_delim c =
  is_ws c || c = '(' || c = ')' || c = '[' || c = ']' || c = '{' || c = '}' || c = '"'

let is_digit c = c >= '0' && c <= '9'

let parse_string (s : string) pos : string * int =
  (* pos sits on the opening quote *)
  let len = String.length s in
  let buf = Buffer.create 16 in
  let rec loop pos =
    if pos >= len then raise (Edn_error "edn: unterminated string");
    match s.[pos] with
    | '"' -> (Buffer.contents buf, pos + 1)
    | '\\' ->
      if pos + 1 >= len then raise (Edn_error "edn: unterminated escape at end of input");
      (match s.[pos + 1] with
       | 'n' -> Buffer.add_char buf '\n'; loop (pos + 2)
       | 't' -> Buffer.add_char buf '\t'; loop (pos + 2)
       | 'r' -> Buffer.add_char buf '\r'; loop (pos + 2)
       | '"' -> Buffer.add_char buf '"'; loop (pos + 2)
       | '\\' -> Buffer.add_char buf '\\'; loop (pos + 2)
       | esc ->
         raise (Edn_error
                  (Printf.sprintf "edn: unsupported escape \\%c at pos %d" esc pos)))
    | c -> Buffer.add_char buf c; loop (pos + 1)
  in
  loop (pos + 1)

let rec parse_value (s : string) (pos : int) : t * int =
  let pos = skip_ws s pos in
  let len = String.length s in
  if pos >= len then raise (Edn_error "edn: unexpected end of input");
  match s.[pos] with
  | '[' ->
    let rec loop pos acc =
      let pos = skip_ws s pos in
      if pos >= len then raise (Edn_error "edn: unterminated vector")
      else if s.[pos] = ']' then (List (List.rev acc), pos + 1)
      else
        let (v, pos) = parse_value s pos in
        loop pos (v :: acc)
    in
    loop (pos + 1) []
  | '(' ->
    let rec loop pos acc =
      let pos = skip_ws s pos in
      if pos >= len then raise (Edn_error "edn: unterminated list")
      else if s.[pos] = ')' then (List (List.rev acc), pos + 1)
      else
        let (v, pos) = parse_value s pos in
        loop pos (v :: acc)
    in
    loop (pos + 1) []
  | '{' -> raise (Edn_error (Printf.sprintf "edn: maps not supported at pos %d" pos))
  | '#' ->
    if pos + 1 < len && s.[pos + 1] = '{' then
      raise (Edn_error (Printf.sprintf "edn: sets not supported at pos %d" pos))
    else
      raise (Edn_error
               (Printf.sprintf "edn: unsupported dispatch #%c at pos %d"
                  (if pos + 1 < len then s.[pos + 1] else '?') pos))
  | '"' ->
    let (str, pos) = parse_string s pos in
    (Str str, pos)
  | _ ->
    let start = pos in
    let rec scan pos =
      if pos < len && not (is_delim s.[pos]) then scan (pos + 1) else pos
    in
    let stop = scan pos in
    if stop = start then
      raise (Edn_error (Printf.sprintf "edn: empty atom at pos %d" start));
    let atom = String.sub s start (stop - start) in
    let v =
      match atom with
      | "nil" -> Nil
      | "true" -> Bool true
      | "false" -> Bool false
      | "_" -> Symbol "_"
      | ":" -> raise (Edn_error (Printf.sprintf "edn: bare ':' at pos %d" start))
      | a when a.[0] = ':' -> Keyword (String.sub a 1 (String.length a - 1))
      | a when is_digit a.[0] || ((a.[0] = '-' || a.[0] = '+') && String.length a > 1 && is_digit a.[1]) ->
        (try Int (Int64.of_string a)
         with Failure _ ->
           (try Float (float_of_string a)
            with Failure _ -> Symbol a))
      | a ->
        (try
           (* float only when plausibly numeric — keeps nan/inf as symbols *)
           let has_digit =
             String.exists (fun c -> is_digit c) a
           in
           if has_digit then Float (float_of_string a) else Symbol a
         with Failure _ -> Symbol a)
    in
    (v, stop)

let read_edn (s : string) : t =
  let (v, pos) = parse_value s 0 in
  let pos = skip_ws s pos in
  if pos <> String.length s then
    raise (Edn_error (Printf.sprintf "edn: trailing content at pos %d" pos));
  v

let read_edn_vector (s : string) : t list =
  let pos = skip_ws s 0 in
  let len = String.length s in
  if pos >= len || s.[pos] <> '[' then
    raise (Edn_error
             (Printf.sprintf "edn: expected tx-data vector, got: %s"
                (if pos < len then String.make 1 s.[pos] else "<eof>")));
  let (elems, _) = parse_value s pos in
  match elems with
  | List xs -> xs
  | _ -> assert false
